File: frontend/src/lib/utils/keepalive.test.ts
import { describe, it, expect } from 'vitest';
import { effectiveKeepalive, keepaliveHint } from './keepalive';

// Таблица обязана совпадать с TestKeepaliveEffective
// (internal/storage/keepalive_test.go): подпись в карточке считает то же
// значение, что бэкенд отправит на прошивку.
describe('effectiveKeepalive', () => {
	const table: Array<[string, number | null]> = [
		['25', 25],
		['25-35', 25],
		[' 25 - 35 ', 25],
		['65535', 65535],
		['', null],
		['0', null],
		[' 0 ', null],
		['0-80', null],
		['65536', null],
		['70000-80000', null],
		['abc', null],
		['-5', null],
		['22-', 22],
		// U+FEFF: JS trim() его снимает, Go strings.TrimSpace — нет. Без
		// явного отказа карточка показала бы «применяется 25 с» там, где
		// бэкенд отвергает правку целиком.
		['\ufeff25', null],
		['25\ufeff', null],
	];

	for (const [raw, want] of table) {
		it(`${JSON.stringify(raw)} → ${want}`, () => {
			expect(effectiveKeepalive(raw)).toBe(want);
		});
	}

	// Одиночное значение приходит из API числом (Keepalive.MarshalJSON).
	it('принимает число и пустые значения из API', () => {
		expect(effectiveKeepalive(25)).toBe(25);
		expect(effectiveKeepalive(0)).toBe(null);
		expect(effectiveKeepalive(null)).toBe(null);
		expect(effectiveKeepalive(undefined)).toBe(null);
	});
});

// Подпись «применяется N с» имеет смысл только там, где прошивка получает
// число: NativeWG. Kernel понимает диапазон целиком, и подпись там врала бы.
describe('keepaliveHint', () => {
	const table: Array<[string | null | undefined, string, number | null]> = [
		['nativewg', '25-35', 25],
		['kernel', '25-35', null],
		['nativewg', '25', null],
		['kernel', '25', null],
		[undefined, '25-35', null],
		['', '25-35', null],
		// Диапазон, который прошивке не отправить: показывать нечего.
		['nativewg', '70000-80000', null],
	];

	for (const [backend, raw, want] of table) {
		it(`${JSON.stringify(backend)} + ${JSON.stringify(raw)} → ${want}`, () => {
			expect(keepaliveHint(backend, raw)).toBe(want);
		});
	}
});

File: internal/storage/keepalive_test.go
package storage

import "testing"

// Effective — единственное правило, по которому keepalive доезжает до
// прошивки: диапазон AWG 3.0 схлопывается в нижнюю границу, а всё, чего
// прошивке не понять, не отправляется вовсе. Тот же список повторяет
// frontend/src/lib/utils/keepalive.ts — расхождение врёт в карточке.
func TestKeepaliveEffective(t *testing.T) {
	for _, tc := range []struct {
		in   Keepalive
		want int
		ok   bool
	}{
		{"25", 25, true},
		{"25-35", 25, true},
		{" 25 - 35 ", 25, true},
		{"65535", 65535, true},
		{"", 0, false},
		{"0", 0, false},
		{" 0 ", 0, false},
		// Нижняя граница 0 — тот же выключенный keepalive, что и "0".
		// ValidateKeepalive такое отвергает, но в записи оно ещё может лежать:
		// keepalive не валидирует ни импорт, ни файлы прошлых версий.
		{"0-80", 0, false},
		// Вне u16: `awg setconf` и NDMS такое не примут, слать нечего.
		{"65536", 0, false},
		{"70000-80000", 0, false},
		{"abc", 0, false},
		{"-5", 0, false},
		// U+FEFF (BOM): strings.TrimSpace пробелом его не считает, а JS trim()
		// снимает. Обе стороны обязаны отвечать одинаково, иначе карточка
		// покажет значение, которое бэкенд отвергнет целиком.
		{"\ufeff25", 0, false},
		{"25\ufeff", 0, false},
		// Формат отсекается ValidateKeepalive до записи; если такое всё же
		// попалось, берём нижнюю границу — она читается однозначно.
		{"22-", 22, true},
	} {
		got, ok := tc.in.Effective()
		if got != tc.want || ok != tc.ok {
			t.Errorf("Keepalive(%q).Effective() = (%d, %v), ждали (%d, %v)", tc.in, got, ok, tc.want, tc.ok)
		}
	}
}
